File: app/src/main/java/com/android/sunshine/app/adapter/ForecastCursorAdapter.java
package com.android.sunshine.app.adapter;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.android.sunshine.app.R;
import com.android.sunshine.app.fragments.ForecastFragmentWeather;
import com.android.sunshine.app.model.OWMWeather;
import com.android.sunshine.app.utils.ApplicationPreferences;
import java.util.List;

public class ForecastCursorAdapter extends RecyclerView.Adapter<ForecastCursorAdapter.ViewHolder> implements OnItemClickHandler {

    private static final int TODAY_VIEW_TYPE = 0;
    private static final int FUTURE_DAY_VIEW_TYPE = 1;

    private Context context;
    private View emptyView;
    private OnAdapterItemClickListener onAdapterItemClickListener;
    private final ApplicationPreferences applicationPreferences;
    private final ItemChoiceManager itemChoiceManager;
    private List<ForecastFragmentWeather> data;

    public ForecastCursorAdapter(final Context context, final View emptyView, final OnAdapterItemClickListener onAdapterItemClickListener,
        final int choiceMode, final ApplicationPreferences applicationPreferences) {
        this.context = context;
        this.emptyView = emptyView;
        this.onAdapterItemClickListener = onAdapterItemClickListener;
        this.applicationPreferences = applicationPreferences;
        itemChoiceManager = new ItemChoiceManager(this);
        itemChoiceManager.setChoiceMode(choiceMode);
    }

    @Override
    public ViewHolder onCreateViewHolder(final ViewGroup viewGroup, final int viewType) {
        if (viewGroup instanceof RecyclerView) {
            int layoutId = -1;
            switch (viewType) {
                case TODAY_VIEW_TYPE:
                    layoutId = R.layout.today_list_item;
                    break;
                case FUTURE_DAY_VIEW_TYPE:
                    layoutId = R.layout.forecast_list_item;
                    break;
                default:
                    break;
            }
            final View view = LayoutInflater.from(context).inflate(layoutId, viewGroup, false);
            return new ViewHolder(view, this);
        } else {
            throw new RuntimeException("Not bound to RecyclerViewSelection");
        }
    }

    @Override
    public void onBindViewHolder(final ViewHolder viewHolder, final int position) {
        ForecastFragmentWeather weather = data.get(position);

        if (getItemViewType(position) == TODAY_VIEW_TYPE) {
            viewHolder.forecastIcon.setImageResource(OWMWeather.getArtResourceForWeatherCondition(weather.getWeatherId()));
            viewHolder.dateWeather.setText(weather.getLongDate());
        } else {
            viewHolder.forecastIcon.setImageResource(OWMWeather.getIconResourceForWeatherCondition(weather.getWeatherId()));
            viewHolder.dateWeather.setText(weather.getShortDate());
        }

        viewHolder.forecastDescription.setText(weather.getDescription());
        viewHolder.max.setText(weather.getMaxTemp());
        viewHolder.min.setText(weather.getMinTemp());
        ViewCompat.setTransitionName(viewHolder.forecastIcon, "iconView" + position);

        itemChoiceManager.onBindViewHolder(viewHolder, position);
    }

    @Override
    public int getItemViewType(final int position) {
        return (position == 0 && applicationPreferences.useTodayLayout()) ? TODAY_VIEW_TYPE : FUTURE_DAY_VIEW_TYPE;
    }

    @Override
    public int getItemCount() {
        return data == null ? 0 : data.size();
    }

    @Override
    public void itemClick(final RecyclerView.ViewHolder viewHolder) {
        int position = viewHolder.getAdapterPosition();
        ForecastFragmentWeather weather = data.get(position);
        onAdapterItemClickListener.onClick(weather.getDateInMillis(), (ViewHolder) viewHolder);
        itemChoiceManager.onClick(viewHolder);
    }

    public void selectView(final RecyclerView.ViewHolder viewHolder) {
        if (viewHolder instanceof ViewHolder) {
            ViewHolder holder = (ViewHolder) viewHolder;
            holder.onClick(holder.itemView);
        }
    }

    public void onRestoreInstanceState(final Bundle savedInstanceState) {
        itemChoiceManager.onRestoreInstanceState(savedInstanceState);
    }

    public void onSaveInstanceState(final Bundle outState) {
        itemChoiceManager.onSaveInstanceState(outState);
    }

    public int getSelectedItemPosition() {
        return itemChoiceManager.getSelectedItemPosition();
    }

    public void setData(final List<ForecastFragmentWeather> data) {
        this.data = data;
        notifyDataSetChanged();
        emptyView.setVisibility(getItemCount() == 0 ? View.VISIBLE : View.GONE);
    }

    public static final class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @Bind(R.id.list_item_icon) public ImageView forecastIcon;
        @Bind(R.id.list_item_date_textview) public TextView dateWeather;
        @Bind(R.id.list_item_forecast_textview) public TextView forecastDescription;
        @Bind(R.id.list_item_high_textview) public TextView max;
        @Bind(R.id.list_item_low_textview) public TextView min;
        private final OnItemClickHandler onItemClickHandler;

        private ViewHolder(final View view, final OnItemClickHandler onItemClickHandler) {
            super(view);
            ButterKnife.bind(this, view);
            this.onItemClickHandler = onItemClickHandler;
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(final View view) {
            onItemClickHandler.itemClick(this);
        }
    }
}

File: app/src/main/kotlin/com/android/sunshine/app/adapter/ForecastCursorAdapter.kt
package com.android.sunshine.app.adapter

import android.content.Context
import android.database.Cursor
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CursorAdapter
import android.widget.ImageView
import android.widget.TextView
import com.android.sunshine.app.R
import com.android.sunshine.app.utils.Utilities

import com.android.sunshine.app.model.WeatherContract.WeatherEntry

public class ForecastCursorAdapter(context: Context, c: Cursor, flags: Int) : CursorAdapter(context, c, flags) {
    public var useTodayLayout: Boolean = false

    override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View {
        val itemViewType = getItemViewType(cursor.getPosition())
        var layoutId = -1
        when (itemViewType) {
            TODAY_VIEW_TYPE -> layoutId = R.layout.today_list_item
            FUTURE_DAY_VIEW_TYPE -> layoutId = R.layout.forecast_list_item
        }
        val view = LayoutInflater.from(context).inflate(layoutId, parent, false)
        val viewHolder = ViewHolder(view)
        view.setTag(viewHolder)
        return view
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val isMetric = Utilities.isMetric(context)
        val weatherId = cursor.getInt(cursor.getColumnIndex(WeatherEntry.COLUMN_WEATHER_ID))
        val weatherDate = cursor.getString(cursor.getColumnIndex(WeatherEntry.COLUMN_DATETEXT))
        val descriptionWeather = cursor.getString(cursor.getColumnIndex(WeatherEntry.COLUMN_SHORT_DESC))
        val maxTemp = cursor.getFloat(cursor.getColumnIndex(WeatherEntry.COLUMN_MAX_TEMP))
        val minTemp = cursor.getFloat(cursor.getColumnIndex(WeatherEntry.COLUMN_MIN_TEMP))

        val viewHolder = view.getTag() as ViewHolder

        viewHolder.dateWeather.setText(Utilities.getFriendlyDay(context, weatherDate))
        viewHolder.forecastDescription.setText(descriptionWeather)
        viewHolder.max.setText(Utilities.formatTemperature(context, maxTemp.toDouble(), isMetric))
        viewHolder.min.setText(Utilities.formatTemperature(context, minTemp.toDouble(), isMetric))

        if (getItemViewType(cursor.getPosition()) == TODAY_VIEW_TYPE) {
            viewHolder.forecastIcon.setImageResource(Utilities.getArtResourceForWeatherCondition(weatherId))
        } else {
            viewHolder.forecastIcon.setImageResource(Utilities.getIconResourceForWeatherCondition(weatherId))
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if ((position == 0 && useTodayLayout)) TODAY_VIEW_TYPE else FUTURE_DAY_VIEW_TYPE
    }

    override fun getViewTypeCount(): Int {
        return 2
    }

    private class ViewHolder public(view: View) {
        public val forecastIcon: ImageView
        public val dateWeather: TextView
        public val forecastDescription: TextView
        public val max: TextView
        public val min: TextView

        {
            forecastIcon = view.findViewById(R.id.list_item_icon) as ImageView
            dateWeather = view.findViewById(R.id.list_item_date) as TextView
            forecastDescription = view.findViewById(R.id.list_item_forecast) as TextView
            max = view.findViewById(R.id.list_item_max) as TextView
            min = view.findViewById(R.id.list_item_min) as TextView
        }
    }

    class object {
        private val TODAY_VIEW_TYPE = 0
        private val FUTURE_DAY_VIEW_TYPE = 1
    }
}